functions/New-DbaCustomError.ps1

```diff
@@ -0,0 +1,171 @@
+            } catch {
+                Stop-Function -Message "Error occurred while establishing connection to $instance" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
+            }
+
+            $languageDetails = $server.Query("SELECT TOP 1 name, alias, msglangid FROM sys.syslanguages WHERE name = '$Language' OR alias = '$Language'")
+
+            if (Test-Bound Language) {
+                if ($null -eq $languageDetails) {
+                    Stop-Function -Message "$server does not have the $Language installed" -Target $instance -Continue
+                }
+            }
+
+            $languageName = $languageDetails.name
+            $languageAlias = $languageDetails.alias
+            $langId = $languageDetails.msglangid
+
+            if ($Pscmdlet.ShouldProcess("Creating new server message with id $MessageID on $server")) {
+                Write-Message -Level Verbose -Message "Creating new server message with id $MessageID on $server"
+                try {
+                    if (Test-Bound Replace) {
+                        $userDefinedMessage = $server.UserDefinedMessages | Where-Object { $_.ID -eq $MessageID -and ($_.Language -in $languageName, $languageAlias -or $_.LanguageID -eq $langId) } # special case for English and British English requires using the langId
+                        $userDefinedMessage.Text = $MessageText
+                        $userDefinedMessage.Severity = $Severity
+                        $userDefinedMessage.Alter()
+                        $newMessages += $userDefinedMessage
+                    } else {
+                        $userDefinedMessage = New-Object -TypeName Microsoft.SqlServer.Management.Smo.UserDefinedMessage
+                        $userDefinedMessage.Parent = $server
+                        $userDefinedMessage.ID = $MessageID
+
+                        if (Test-Bound Language) {
+                            $userDefinedMessage.Language = $Language
+                        } else {
+                            $userDefinedMessage.Language = ($server.Query("SELECT syslang.name FROM sys.syslanguages syslang JOIN sys.configurations config ON syslang.langid = config.value_in_use AND config.name = 'default language'")).name
+                        }
+
+                        $userDefinedMessage.Severity = $Severity
+                        $userDefinedMessage.Text = $MessageText
+
+                        if (Test-Bound WithLog) {
+                            $userDefinedMessage.IsLogged = $true
+                        }
+
+                        $userDefinedMessage.Create()
+
+                        # pull the new message object from the server to get all properties refreshed (the $userDefinedMessage.Refresh() method does not work as expected)
+                        $newMessages += $server.UserDefinedMessages | Where-Object { $_.ID -eq $MessageID -and $_.Language -eq $userDefinedMessage.Language }
```

Comment:
We prefer to not collect objects before returning. The use of `+=` is a performance hit in large environments. The preferred method is to output the data as soon as you can.

In other commands, you can find the use of `Select-DefaultView` where we handle a set of properties out by default when the object in SMO outputs a large number of properties.

Reply:
ok, sounds good